core/templates/dev/head/exploration_editor/EditorServices.js

```diff
@@ -1952,3 +2012,283 @@ oppia.factory('explorationWarningsService', [
     };
   }
 ]);
+
+oppia.factory('lostChangesService', ['utilsService', function(utilsService) {
+  var CMD_ADD_STATE = 'add_state';
+  var CMD_RENAME_STATE = 'rename_state';
+  var CMD_DELETE_STATE = 'delete_state';
+  var CMD_EDIT_STATE_PROPERTY = 'edit_state_property';
+
+  var getRulesList = function(value) {
+    var rulesList = [];
+    value.rule_specs.forEach(function(ruleSpec) {
+      var ruleElm = angular.element('<div></div>');
+      ruleElm.append('<p>Type: ' + ruleSpec.rule_type + '</p>');
+      ruleElm.append(
+        '<p>Value: ' +
+          (
+            Object.keys(ruleSpec.inputs).map(function(input) {
+              return ruleSpec.inputs[input];
+            })
+          ).toString() +
+          '</p>');
+      rulesList.push(ruleElm);
+    });
+
+    return rulesList;
+  };
+
+  var getStatePropertyValue = function(
+    statePropertyValue) {
+    return angular.isArray(statePropertyValue) ?
+      statePropertyValue[statePropertyValue.length - 1] : statePropertyValue;
+  };
+
+  var isEmpty = utilsService.isEmpty;
+
+  var getRelativeChangeToGroups = function(changeObject) {
+    var newValue = changeObject.new_value;
+    var oldValue = changeObject.old_value;
+    var result = '';
+
+    if (angular.isArray(newValue) && angular.isArray(oldValue)) {
+      result = (newValue.length > oldValue.length) ?
+        'added' : (newValue.length === oldValue.length) ?
+        'edited' : 'deleted';
+    } else {
+      if (!isEmpty(oldValue)) {
+        if (!isEmpty(newValue)) {
+          result = 'edited';
+        } else {
+          result = 'deleted';
+        }
+      } else if (!isEmpty(newValue)) {
+        result = 'added';
+      }
+    }
+    return result;
+  };
+
+  var makeHumanReadable = function(lostChanges) {
+    var outerHtml = angular.element('<ul></ul>');
+    var statesWiseEditsMapping = {};
+
+    lostChanges.forEach(function(lostChange) {
+      var lostChangeEl = angular.element('<li></li>');
+      var lostChangeValue;
+      switch (lostChange.cmd) {
+        case CMD_ADD_STATE:
+          lostChangeValue = 'Added state: ' + lostChange.state_name;
+          break;
+        case CMD_RENAME_STATE:
+          lostChangeValue = ('Renamed state: ' + lostChange.old_state_name +
+                             ' to ' + lostChange.new_state_name);
+          break;
+        case CMD_DELETE_STATE:
+          lostChangeValue = 'Deleted state: ' + lostChange.state_name;
+          break;
+        case CMD_EDIT_STATE_PROPERTY:
+          var newValue = getStatePropertyValue(lostChange.new_value);
+          var oldValue = getStatePropertyValue(lostChange.old_value);
+          var stateName = lostChange.state_name;
+
```

Comment:
I think it would be clearer if you create an entry in statesWiseEditMappings (if necessary) at this point. Then within each case you can just add an appropriate element to statesWiseEditMappings and you're done. Currently the reader has to worry about what exactly is going to happen to the lostChangeValue and the lostChangeEl at the end.